### PR TITLE
Don't drag function signatures along function item types.

### PR DESCRIPTION
**Pull request by eddyb**
This PR separates the signature of a function from the "function item type" (`TyFnDef`), leaving only the `DefId` and parameter `Substs`, making them even more like (captureless) closure types.

The motivation for this change is reducing typesystem complexity, and its consequences:
* operating on the signature instead of just the parameters was less efficient
  * specifically, signatures can easily add several levels of depth on top of the parameter types
  * and the signatured were always substituted and normalized, so typically even more complex
* it was *the only* type that was *both* nominal (identity) and structural (signature)
  * harder to model in Chalk than either a purely nominal or structural type
  * subtyping worked on the signature but parameters were always invariant
  * call type-checking was transforming signatures but keeping the nominal half intact
  * the signature could therefore get out of sync during type inference in several ways

That last point comes with a `[breaking-change]`, because functions with `'static` in their return types will now *not* be as usable as if they were using lifetime parameters instead:
```rust
// Will cause lifetime mismatch in main after this PR.
fn bar() -> &'static str { "bar" }
// Will continue to work fine, as every use can choose its own lifetime.
fn bar<'a>() -> &'a str { "bar" }

fn main() {
    let s = String::from("foo");
    Some(&s[..]).unwrap_or_else(bar);
}
```

r? @nikomatsakis

### CODE REVIEW

File: src/librustc/ty/structural_impls.rs
Comment (petrochenkov) [inline src/librustc/ty/structural_impls.rs]: Is it okay that signatures are not folded anymore? (Is it possible to fold something kept in the tables at all?)
It would be really nice if they were at least still *visited* by default.

Comment (eddyb) [inline src/librustc/ty/structural_impls.rs]: The whole point (which I'll try to describe once I'm sure there's no fatal flaw with the approach) is to reduce the amount of work done for these types. *Anything* that relied on seeing the signature has to visit it manually. It's not a property of the type and it was never meant to be, it's just one way to use values of that type (i.e. calling them).

**Comment by Mark-Simulacrum** [comment]
@bors try

This will get us artifacts for cargobomb.

**Comment by Mark-Simulacrum** [comment]
Don't retry until the previous try finishes one way or another.

**Comment by Mark-Simulacrum** [comment]
Oh right

@bors retry

**Comment by eddyb** [comment]
Started crater run (fingers crossed it's fixed now).

**Comment by tomprince** [comment]
I've started a cargobomb run of this.

**Review by nikomatsakis — APPROVED** [review]
So, I think these changes are good, and are probably the right thing to do. Two things give me mild discomfort:

(a) the fact that subtyping may cause impls to change. 
(b) the fact that we default to invariant for "unused" types. 

But both seem ok. I guess (b) is a conservative choice.

**Comment by nikomatsakis** [comment]
I feel like I'm ready to r=me this PR, but I feel like I would like to at least *briefly* discuss it with all of @rust-lang/lang -- it feels like a change we should be aware of.

Context:

In the current rustc compiler implementation, the "fn item type" for a function is implemented in a way that carries its signature along with it, and it allows the signature to vary *independently of the original input types*. This is weird. We'd prefer to model fn item types roughly like this:

```rust
fn foo<'a, A, B>(x: &'a A) -> &'a B { ... }

struct Foo<'a, A> {
    data: PhantomData<fn(&'a A) -> &'a B>
}
```

This is roughly what @eddyb implemented, but with the following caveat:

- If some (eager) lifetime or type parameter is not used in the fn signature, then it becomes **invariant**.

I also noticed an orthogonal, pre-existing situation around variance that makes me sad. In particular, `fn(&u8)` (where the region is bound) can have a distinct impl for some trait from `fn(&'a u8)` (where the region is free). But also `fn(&u8)` is a subtype of `fn(&'a u8)`. So that means that if you `Foo<fn(&u8)>` you can upcast it to `Foo<fn(&'a u8)>` (presuming `Foo` is covariant w/r/t its type argument). If you then have `struct Foo<T: Trait>`, that means subtyping might change the impl that executes. I don't believe this is ever *unsound* (e.g., if we are projecting types out, we'll notice *that* elsewhere), but it's weird. My preferred fix here though is to remove the rules that allow you to have distinct trait impls for bound-and-unbound regions, if we can get away with it.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I propose that we merge this PR. A description of its contents and changes [can be found here](https://github.com/rust-lang/rust/pull/42417#issuecomment-307501286).

File: src/librustc/infer/mod.rs
Comment (eddyb) [inline src/librustc/infer/mod.rs]: I believe someone also suggested `fn_sig_of` or just `sig_of`. But the latter doesn't fit with the type name.

**Comment by tomprince** [comment]
The cargobomb results are [here](http://cargobomb-reports.s3.amazonaws.com/pr-42417/index.html).

**Comment by eddyb** [comment]
The [only regression](https://cargobomb-reports.s3.amazonaws.com/pr-42417/aa4570afb47abc5280a3360cd347910ab9c6ede4/reg/cargo-fuzz-0.4.2/log.txt) comes from code that relies on signature subtyping, and can be reduced to:
```rust
fn bar() -> &'static str { "bar" }

fn main() {
    let s = String::from("foo");
    Some(&s[..]).unwrap_or_else(bar);
}
```
Before, we'd subtype the signatures and have `-> &'_ str` which would unify with the type of `&s`.
But now it's `'static`, and because `Fn` bounds *don't* have variance, `&s` is required to be `'static`.

**Comment by eddyb** [comment]
One thing we could do here is to generalize (produce a supertype of) the return type, when projecting out `FnOnce::Output` from a function item type.

I don't think the same makes sense for the arguments, as for the function definition to require *less* than the bound, but that naturally results in lifetime parameters *which have variance*, so they would naturally handle all the subtyping - it's really *only* `'static` in the return type that changed.

**Comment by nikomatsakis** [comment]
So @eddyb and I chatted a bit on IRC about what precisely he is proposing. Roughly speaking the idea would be, when you project `Output` from a fn item, you would get "some supertype" of the declared return type. That is sort of equivalent to an impl like this:

```rust
// Given:
//     fn foo<A>() -> A
// Produce:
impl<R, A> FnOnce for Foo<A>
where A: Subtype<R>
{
    type Output = R;
}
```

if `Subtype` were some special built-in trait. The purpose of this would be to [avoid this regression](https://github.com/rust-lang/rust/pull/42417#issuecomment-307548774). 

@eddyb, does this sound about right?

**Comment by eddyb** [comment]
@nikomatsakis That seems like an accurate description, although *that example* already gets the behavior in question because  `Foo` is covariant over `A` for all intents and purposes, AFAIK.

**Comment by nikomatsakis** [comment]
@eddyb 

> that example already gets the behavior in question because Foo is covariant over A for all intents and purposes, AFAIK.

Yes. We could change it to `fn foo<A>(A) -> A` to see the new behavior I guess.

**Comment by arielb1** [comment]
That also something I've thought about (actually, I've thought of doing subtyping on the entire fn type, which differs on "is `fn(&u32): Fn(&'static u32)`"). I can't think of any problems with that idea - that's still a "trivial shim", but I'm not sure I want to introduce subtyping to trait selection in such a "non-reified" way.

**Comment by nikomatsakis** [comment]
@arielb1 can you elaborate on what you mean by "non-reified"? I imagine if we implemented this, we'd probably use the `Subtype` obligations that we now have to transmit that information.

**Comment by nikomatsakis** [comment]
/me grumbles about that "close and comment" button

**Comment by pnkfelix** [comment]
@eddyb @nikomatsakis are you two talking about making the change under discussion *before* merging this PR? Or as some follow-up task? (I *assume* you mean to make it before merging, as to avoid the aforementioned regression, but I just wanted to confirm, since I would be willing to check off my box if that assumption is true...)

**Comment by eddyb** [comment]
@pnkfelix If this PR is accepted I'd make the `Output` subtyping change before merging, yes.

**Comment by nikomatsakis** [comment]
@pnkfelix yes I meant before merging.

@arielb1 I agree it's a little magical; I sort of expect that sooner-or-later we will wind up with a `Coerce` trait (that users probably cannot implement...) to represent "coercible values" (a superset of subtyping), and one could imagine then users using this trait to do the same thing. I *suppose* we could expose a magic `Subtype` trait too, but it's not clear that it's worth forcing this distinction onto users. It maybe, though. (People often talk about "transmutable", which is certainly strongly related.)

**Comment by nikomatsakis** [comment]
I took the liberty of tagging @aturon since he's on PTO. Although, he *does* get back next Monday... well, I'll chat with him about it.

**Comment by Mark-Simulacrum** [comment]
@eddyb I believe this may be ready to go other than a description. I .. don't really know if we need to wait for FCP -- seems that team has given approval, let's land as soon as we get a description.

**Comment by eddyb** [comment]
@Mark-Simulacrum I also have to implementing a subtyping mechanism for `FnOnce::Output`.

**Comment by carols10cents** [comment]
> I believe this may be ready to go other than a description. I .. don't really know if we need to wait for FCP -- seems that team has given approval, let's land as soon as we get a description.

@Mark-Simulacrum 
1rj00n

**Comment by eddyb** [comment]
I'm not sure we can do the subtyping thing, imagine this:
```rust
struct FnAndValue<F: Fn()> {
    f: F,
    value: F::Output
}

fn launder<T>(x: &T) -> &'static T {
    let mut laundromat = FnAndValue {
        f: (|| None) as fn() -> Option<&'static T>,
        value: None
    };
    laundromat.value = Some(x);
    laundromat.value
}
```
If we normalize to a subtype every time, the *guarantees* of the language that associated types are *fully* determined by the input parameters, are lost, and the same type can be different now.
This is like (unsound) late-bound lifetimes in return types, which is what I proposed `'static` be here.

**Comment by eddyb** [comment]
Much fuss for nothing: I was trying to find the code in `cargo-fuzz` again and I realized that *while* `0.4.2` doesn't compile, `0.4.3` does just fine (it now calls the function instead of passing it).
*Also*, `0.4.1` compiles. So it's just one patch version, which could be yanked.
cc @frewsxcv @nagisa @Manishearth

**Comment by jethrogb** [comment]
Can we add a better error message when people encounter this breaking change? You now just get a generic borrow check error.

**Comment by eddyb** [comment]
@jethrogb Are you talking about the example itself or have you seen this in the wild *in another form*? And no, you can't detect this without *reintroducing* the complex representation removed here.

**Comment by jethrogb** [comment]
@eddyb I just came here from the draft release notes. I haven't actually encountered any broken code myself. I do think it's fairly common for people to write `&'static str` as a function return type, so I think this error is likely to come up during development. The error message doesn't imply "you can fix this with a reborrow" which will likely be the case.

I don't exactly understand how this change makes detecting this harder but maybe some kind of heuristic (instead of a guaranteed detection) is good enough. Alternatively, maybe we can make a documentation change somewhere.

**Comment by eddyb** [comment]
You only get an error if you pass such a function to another function that usually takes a closure.
Reborrows on immutable references don't usually do anything.

**Comment by jethrogb** [comment]
I meant this is a way to fix the code above:
```rust
    Some(&s[..]).unwrap_or_else(||&*bar());
```

**Comment by eddyb** [comment]
Oh and the canonical fix is to make the `'static` into a lifetime parameter.
In your snippet you can remove the reborrow. `|| bar()` is the fix there, the reborrow does nothing more.

**Comment by jethrogb** [comment]
> Oh and the canonical fix is to make the `'static` into a lifetime parameter.

Sure but the user doesn't always have a choice in that regard. Popular crates use `&'static str` in their return types:
https://docs.rs/toml/0.4/toml/value/enum.Value.html#method.type_str
http://docs.diesel.rs/diesel/query_source/trait.Column.html#tymethod.name

**Comment by eddyb** [comment]
Fair enough, the second best choice is introducing a closure (like you did, if you ignore the reborrow).

**Pull request merged**
